src/shared/cloudformation/templateSymbolResolver.ts

```diff
@@ -27,45 +22,106 @@ export interface TemplateSymbolProvider {
+                    // We want the resource name of the Function node (not Api,
+                    // that node is always named "Events".)
+                    name: funSymbol.name,
+                    range: o.range,
+                    kind: kind,
+                }))
+            )
+        }
+        return apiSymbols
     }
 
-    private isFunctionResource(symbol: vscode.DocumentSymbol) {
-        const typeSymbol = _(symbol.children).find({ name: 'Type', kind: vscode.SymbolKind.String })
+    /**
+     * Searches the tree and returns all matching nodes.
+     */
+    private findDescendants(
+        symbols: vscode.DocumentSymbol[],
+        name: string,
+        kind: vscode.SymbolKind
+    ): vscode.DocumentSymbol[] {
+        const found = symbols.filter(v => v.name === name && v.kind === kind)
+        for (let s of symbols) {
+            found.push(...this.findDescendants(s.children, name, kind))
+        }
+        return found
+    }
 
+    private isKind(kind: 'function' | 'api', symbol: vscode.DocumentSymbol) {
+        // Example:
+        // ------------------------------------
+        // Resources:
+        //   ImplicitApiFunction:
+        //     Type: AWS::Serverless::Function
+        //     Properties:
+        //       ...
+        //       Events:             \
+        //         GetHtml:           \
+        //           Type: Api         \ presence of this means
+        //           Properties:       / API instead of FUNCTION
+        //             Path: /        /
+        //             Method: get   /
+        const typeSymbol = this.findDescendants(symbol.children, 'Type', vscode.SymbolKind.String)[0]
+        const wantType =
+            kind === 'function' ? CloudFormation.SERVERLESS_FUNCTION_TYPE : CloudFormation.SERVERLESS_API_TYPE
         if (!typeSymbol) {
             return false
         }
-
         const parsedSymbol = safeLoad(this.symbolProvider.getText(typeSymbol, this.document)) as { Type: string }
-
-        return parsedSymbol.Type === CloudFormation.SERVERLESS_FUNCTION_TYPE
+        return parsedSymbol.Type === wantType
     }
 }
 
-class DefaultSymbolProvider implements TemplateSymbolProvider {
-    public async getSymbols(document: vscode.TextDocument): Promise<vscode.DocumentSymbol[]> {
-        const symbols = await vscode.commands.executeCommand<vscode.DocumentSymbol[]>(
-            'vscode.executeDocumentSymbolProvider',
-            document.uri
-        )
+export class TemplateSymbolProvider {
+    public async getSymbols(document: vscode.TextDocument, waitForSymbols: boolean): Promise<vscode.DocumentSymbol[]> {
+        for (let i = 0; i < 30; i++) {
+            const symbols = await vscode.commands.executeCommand<vscode.DocumentSymbol[]>(
+                'vscode.executeDocumentSymbolProvider',
+                document.uri
+            )
+            if (symbols !== undefined || !waitForSymbols) {
+                return symbols || []
```

Comment:
Why do we need `||`? we already know it's not undefined and if it's empty it's empty

Reply:
It may be undefined if caller passed `waitForSymbols=false`.

Reply:
style question, why `||` over `??`

Reply:
`||` is just defensive in case it's otherwise falsy. Doesn't matter but I'll change it.